common/src/main/java/com/wynntils/core/json/JsonManager.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.core.json;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonParser;
+import com.google.gson.stream.JsonReader;
+import com.wynntils.core.WynntilsMod;
+import com.wynntils.core.components.Manager;
+import com.wynntils.utils.FileUtils;
+import com.wynntils.utils.colors.CustomColor;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.lang.reflect.Field;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.apache.commons.lang3.reflect.FieldUtils;
+
+public final class JsonManager extends Manager {
+    public static final Gson GSON = new GsonBuilder()
+            .registerTypeAdapter(CustomColor.class, new CustomColor.CustomColorSerializer())
+            .setPrettyPrinting()
+            .serializeNulls()
+            .create();
+
+    public JsonManager() {
+        super(List.of());
+    }
+
+    public Type findFieldTypeOverride(Object parent, Field field) {
+        Optional<Field> typeField = Arrays.stream(
+                        FieldUtils.getFieldsWithAnnotation(parent.getClass(), TypeOverride.class))
+                .filter(f -> f.getType() == Type.class && f.getName().equals(field.getName() + "Type"))
+                .findFirst();
+
+        if (typeField.isPresent()) {
+            try {
+                return (Type) FieldUtils.readField(typeField.get(), parent, true);
+            } catch (IllegalAccessException e) {
+                WynntilsMod.error("Unable to get field " + typeField.get().getName(), e);
+            }
+        }
+
+        return null;
+    }
+
+    public Object deepCopy(Object value, Type fieldType) {
+        return GSON.fromJson(GSON.toJson(value), fieldType);
+    }
+
+    public void savePreciousJson(File jsonFile, JsonObject jsonObject) {
+        // create file if necessary
+        if (!jsonFile.exists()) {
+            FileUtils.createNewFile(jsonFile);
+        }
+
+        // FIXME: Make backup first!
```

Comment:
Did we lose the ability to backup our old config if something went wrong? Is this a future improvement or a thing  you wanted to do in this PR?

Reply:
I now read the code below. I am not fully sure what this comment means though. I assume we want to not corrupt our config and/or have a way to revert if something goes wrong here? (in which case I am not sure I can think of a legit case of happening)

Reply:
It is a future improvement. I wanted to do that as part of this PR but forgot about it. It's rather trivial (i just want to rename our old json file to .bak before writing a new one), so I can fix it now.